source/pkg/adapter/adapter.go

```diff
@@ -92,30 +105,46 @@ func (a *Adapter) Start(ctx context.Context) error {
 			a.logger.Info("Reusing consumer group", zap.String("group", groupName))
 		} else {
 			a.logger.Info("Creating consumer group", zap.String("group", groupName))
-			_, err := conn.Do("XGROUP", "CREATE", a.config.Stream, groupName, "$")
+			_, err := conn.Do("XGROUP", "CREATE", streamName, groupName, "$")
 			if err != nil {
 				a.logger.Error("Cannot create consumer group", zap.Error(err))
 				return err
 			}
 		}
 
 	}
-	conn.Close()
 
 	a.logger.Info("Number of consumers from config:", zap.Int("NumConsumers", a.config.NumConsumers))
 
 	for i := 0; i < a.config.NumConsumers; i++ {
-		go func(j int) {
+		waitGroup.Add(1)
+
+		go func(shutdownChannel chan struct{}, wg *sync.WaitGroup, j int) {
+			defer wg.Done()
+
 			conn, _ := pool.Dial()
 
 			consumerName := fmt.Sprintf("%s-%d", groupName, j)
 			xreadID := "0" //Initial ID to read pending messages
 			a.logger.Info("Listening for messages", zap.String("consumerName", consumerName))
 
 			for {
-				//XREAD below reads all the pending messages when xreadID=="0" and new messages when xreadID==">"
-				reply, err := conn.Do("XREADGROUP", "GROUP", groupName, consumerName, "COUNT", 1, "BLOCK", 0, "STREAMS", a.config.Stream, xreadID)
+				//TODO: Check if statefulset's adapter pod is killed, and do not read more messages (to finish up work)
+				select {
+				case <-shutdownChannel:
+					a.logger.Info("Shutdown consumer", zap.String("consumerName", consumerName))
+
+					_, err := conn.Do("XGROUP", "DELCONSUMER", streamName, groupName, consumerName)
```

Comment:
what happens to the pending messages? Maybe there are none?

Reply:
I don't think we can be sure of that. So better to check for pending maybe